huge bar fix

Ayro97 opened the pull request:
fix for https://github.com/Zidras/DBM-Warmane/issues/252

3 key changes: 

**1. Correcting When the Bar Becomes Huge**
The Problem: Bars with variance were becoming huge too early. The code was checking the displayed text (time until variance starts) against the enlargeTime, not the bar's actual timer.
The Solution: Change the enlargement condition in `barPrototype:Update` to always use the bar's actual timer `(timerValue)`.

**2. Correcting the Huge Bar's Fill Level**
The Problem: The huge bar's visual fullness was still a percentage of its original total duration, making it look half-empty when it should have been starting a new countdown.
The Solution: In `barPrototype:Update`, for huge "NoAnim" bars, changed the denominator in the `bar:SetValue()` calculation from the original `totaltimeValue` to `enlargeTime`.

**3. Correcting the Huge Bar's Variance Overlay**
The Problem: The shaded variance overlay was too small on huge bars because its width was also being calculated as a percentage of the original, longer duration.
The Solution: Modified the `barPrototype:SetVariance` function. Added a check to see if the bar was huge and using the "NoAnim" style. If so, change the denominator for the overlay's width calculation from `self.totalTime` to `enlargeTime`.

Ayro97 committed:
huge bar fix

Ayro97 committed:
Update DBT.lua

remove whitespace

Zidras commented on DBM-StatusBarTimers/DBT.lua:
```diff
@@ -334,111 +334,115 @@ do
-			timer = varianceMinTimer or varianceMaxTimer -- varianceMaxTimer here could be just normal number timer, so check for varianceMinTimer, which only exists if it's a variant timer
-		end
-		if not timer or (self.numBars >= 15 and not isDummy) then
+function DBT:CreateBar(timer, id, icon, huge, small, color, isDummy, colorType, inlineIcon, keep, fade, countdown, countdownMax)
```
This is wrong indentation. Revert to what it was before, with 1 line break in between functions

Zidras commented on DBM-StatusBarTimers/DBT.lua:
```diff
@@ -334,111 +334,115 @@ do
+		newBar.minTimer = varianceMinTimer or nil
+		newBar.varianceDuration = varianceDuration or 0
+		newBar.hasVariance = varianceMinTimer and true or false
+		-- FIX: Reset totalTime BEFORE SetTimer to ensure proper variance handling
+		newBar.totalTime = timer
+		newBar:SetTimer(timer) -- This can kill the timer and the timer methods don't like dead timers
```
This call is inside SetTimer. I don't understand the need for this code nor what it tries to fix

Zidras commented on DBM-StatusBarTimers/DBT.lua:
```diff
@@ -334,111 +334,115 @@ do
-		-- Most efficient place to block it, nil colorType instead of checking option every update
-		if not self.Options.ColorByType then
-			colorType = nil
+		-- FIX: Ensure elapsed time is properly reset for variance timers
+		-- Reset elapsed before setting it to 0 to prevent percentage calculation issues
+		newBar.elapsed = 0
+		newBar:SetElapsed(0)
```
What is this new element you're introducing `.elapsed` ? I don't see it being used anywhere

Zidras commented on DBM-StatusBarTimers/DBT.lua:
```diff
@@ -829,7 +833,6 @@ function barPrototype:SetIcon(icon)
 
 function barPrototype:SetColor(color)
-	-- Fix to allow colors not require the table keys
 	if color[1] and not color.r then
```
Why remove this comment?

Zidras commented on DBM-StatusBarTimers/DBT.lua:
```diff
@@ -848,14 +851,25 @@ function barPrototype:SetVariance()
-		varianceTex:SetWidth(varianceWidth)
+		local varianceWidth
+		local isEnlarged = self.enlarged and not self.paused
+		local barOptions = DBT.Options
```
You cached DBT.Options but only used it once.

Zidras commented on DBM-StatusBarTimers/DBT.lua:
```diff
@@ -848,14 +851,25 @@ function barPrototype:SetVariance()
+		end
 
-		-- change SetPoints based on fillUpBars
+		varianceTex:SetWidth(varianceWidth)
```
Why remove the comment?

Zidras commented on DBM-StatusBarTimers/DBT.lua:
```diff
@@ -974,22 +988,20 @@ function barPrototype:Update(elapsed)
-			end
-		end
-		timer:SetText(stringFromTimer(timerCorrectedNegative))
+if fillUpBars then
```
Wrong indentation

Zidras commented on DBM-StatusBarTimers/DBT.lua:
```diff
@@ -974,22 +988,20 @@ function barPrototype:Update(elapsed)
-		end
-		timer:SetText(stringFromTimer(timerCorrectedNegative))
+if fillUpBars then
+    if currentStyle == "NoAnim" and timerValue <= enlargeTime and not enlargeHack then
```
Why remove comments?

Zidras commented on DBM-StatusBarTimers/DBT.lua:
```diff
@@ -1073,7 +1085,7 @@ function barPrototype:Update(elapsed)
 		DBT:UpdateBars(true)
 	end
-	if not paused and ((barOptions.VarianceEnabled and timerLowestValueFromVariance or timerValue) <= enlargeTime) and not self.small and not isEnlarged and isMoving ~= "enlarge" and enlargeEnabled then
+	if not paused and timerValue <= enlargeTime and not self.small and not isEnlarged and isMoving ~= "enlarge" and enlargeEnabled then
```
IIRC, this was intended. The bar starts enlarging at the lowest possible time, not the highest possible.

Zidras reviewed (changes requested):
I have reviewed the PR purely on submitted code, I haven't tested anything, so I might not be grasping your intentions.
In any case, it seems there are extra code unrelated to the initial bug report. Not only that, some even seem wrong.

Ayro97 committed:
Update DBT.lua

https://github.com/Zidras/DBM-Warmane/pull/257/files/75d8cad6055897a7af800e878b07c986a042ea22

Ayro97 committed:
Update DBT.lua
